[identity] Latest API updates for Azure pipelines credential (#29854)
Azure/azure-sdk-for-js: sdk/identity/identity/TROUBLESHOOTING.md

## Troubleshoot environment credential authentication issues

### CredentialUnavailableError

| Error Message                                  | Description                                              | Mitigation                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

> | ---------------------------------------------- | -------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

Reviewer: nit: no need add more than three dashes for each column definition [docs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-tables)

Author: Done

## Troubleshoot `AzurePipelinesCredential` authentication issues

| Error Message                                                                                                                                                                                                                                                                                                                                                            | Description                                                                                                                                                                       | Mitigation                                                                                                                                                                                                                                                                                                                                                                                                             |
| --- | ---| --- |

> | AADSTS900023: Specified tenant identifier '<some tenant Id>' is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                                   | The Microsoft Entra tenant ID passed to the credential is invalid.                                                                                                                | Verify the tenant id is valid. If the service connection FIC was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection FIC is configured via an app registration, the tenant should be the one in which the app registration is registered.                                                                                  |

Reviewer 2: ```suggestion
| AADSTS900023: Specified tenant identifier '<some tenant Id>' is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                                   | The Microsoft Entra tenant ID passed to the credential is invalid.                                                                                                                | Verify the tenant ID is valid. If the service connection federated identity credential (FIC) was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection FIC is configured via an app registration, the tenant should be the one in which the app registration is registered.                                                                                  |
```

> | AADSTS900023: Specified tenant identifier '<some tenant Id>' is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                                   | The Microsoft Entra tenant ID passed to the credential is invalid.                                                                                                                | Verify the tenant ID is valid. If the service connection federated identity credential (FIC) was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection FIC is configured via an app registration, the tenant should be the one in which the app registration is registered.                                                                                  |

Reviewer 3: Let's backtick `<some tenant ID>` as right now it's being rendered as an HTML tag and doesn't appear.

```suggestion
| AADSTS900023: Specified tenant identifier `<some tenant ID>` is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                                   | The Microsoft Entra tenant ID passed to the credential is invalid.                                                                                                                | Verify the tenant ID is valid. If the service connection federated identity credential (FIC) was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection FIC is configured via an app registration, the tenant should be the one in which the app registration is registered.                                                                                  |
```

> | AADSTS900023: Specified tenant identifier '975f013f-7f24-47e8-a7d3-abc4752bf34' is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                | The tenant id passed to the credential is invalid                                                                                                                                 | Verify the tenant id is valid. If the service connection was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection is configured via a Service Principal, the tenant should be the one in which the Service Principal is registered.                                                                             |

Reviewer 4: UAMIs each have a Service Principal. App Registrations have Enterprise Apps, each of which has a Service Principal. Do not use Service Principal as a name for App Registrations, it's not correct.

Author: done

> | AADSTS900023: Specified tenant identifier '975f013f-7f24-47e8-a7d3-abc4752bf34' is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                | The tenant id passed to the credential is invalid                                                                                                                                 | Verify the tenant id is valid. If the service connection was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection is configured via a Service Principal, the tenant should be the one in which the Service Principal is registered.                                                                             |

Reviewer: ```suggestion
| AADSTS900023: Specified tenant identifier '<some tenant Id>' is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                | The tenant id passed to the credential is invalid                                                                                                                                 | Verify the tenant id is valid. If the service connection was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection is configured via a Service Principal, the tenant should be the one in which the Service Principal is registered.                                                                             |
```

> | AADSTS900023: Specified tenant identifier '975f013f-7f24-47e8-a7d3-abc4752bf34' is neither a valid DNS name, nor a valid external domain.                                                                                                                                                                                                                                | The tenant id passed to the credential is invalid                                                                                                                                 | Verify the tenant id is valid. If the service connection was configured via a user-assigned managed identity, the tenant will be the one in which managed identity was registered. If the service connection is configured via a Service Principal, the tenant should be the one in which the Service Principal is registered.                                                                             |

Author: done

> | No service connection found with identifier <GUID>                                                                                                                                                                                                                                                                                                                       | The service connection ID provided is incorrect.                                                                                                                                  | Verify the  `serviceConnectionId` argument provided. This parameter refers to the `resourceId` of the Azure Service Connection. It can also be found in the query string of the respective Service Connection's configuration page in Azure Pipelines. More information about service connections can be found [here](https://learn.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml)       |

Reviewer 2: ```suggestion
| No service connection found with identifier <GUID>                                                                                                                                                                                                                                                                                                                       | The service connection ID provided is incorrect.                                                                                                                                  | Verify the `serviceConnectionId` argument provided. This parameter refers to the `resourceId` of the Azure service connection. It can also be found in the query string of the respective service connection's configuration page in Azure Pipelines. More information about service connections can be found [here](https://learn.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml).       |
```

> | No service connection found with identifier <GUID>                                                                                                                                                                                                                                                                                                                       | The service connection ID provided is incorrect.                                                                                                                                  | Verify the `serviceConnectionId` argument provided. This parameter refers to the `resourceId` of the Azure service connection. It can also be found in the query string of the respective service connection's configuration page in Azure Pipelines. More information about service connections can be found [here](https://learn.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml).       |

Reviewer 3: Let's backtick `<GUID>` as right now it's being rendered as an HTML tag and doesn't appear.

```suggestion
| No service connection found with identifier `<GUID>`                                                                                                                                                                                                                                                                                                                       | The service connection ID provided is incorrect.                                                                                                                                  | Verify the `serviceConnectionId` argument provided. This parameter refers to the `resourceId` of the Azure service connection. It can also be found in the query string of the respective service connection's configuration page in Azure Pipelines. More information about service connections can be found [here](https://learn.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml).       |
```

> | No service connection found with identifier <GUID>                                                                                                                                                                                                                                                                                                                       | The service connection id provided is incorrect.                                                                                                                                  | Verify the serviceConnectionId provided. This parameter refers to the `resourceId` of the Azure Service Connection. It can also be found in the queryString of the respective Service Connection's configuration page in Azure Devops.                                                                                                                                                                     |

Reviewer: nit: 
```suggestion
| No service connection found with identifier <GUID>                                                                                                                                                                                                                                                                                                                       | The service connection id provided is incorrect.                                                                                                                                  | Verify the serviceConnectionId provided. This parameter refers to the `resourceId` of the Azure Service Connection. It can also be found in the query string of the respective Service Connection's configuration page in Azure DevOps. More information about service connections can be found [here](https://learn.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml) |
```

> | No service connection found with identifier <GUID>                                                                                                                                                                                                                                                                                                                       | The service connection id provided is incorrect.                                                                                                                                  | Verify the serviceConnectionId provided. This parameter refers to the `resourceId` of the Azure Service Connection. It can also be found in the queryString of the respective Service Connection's configuration page in Azure Devops.                                                                                                                                                                     |

Author: done

| AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = Object moved to here. Status Code: 302.                                                                                                                                                                                                                        | The system access token seems to be malformed when passing in as a parameter to the credential.                                                                                   | `System.AccessToken` is a required system variable in the Azure Pipelines task and should be provided in the pipeline task, [as mentioned in the docs](https://learn.microsoft.com/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#systemaccesstoken). Verify that the system access token value provided is the predefined variable in Azure Pipelines and isn't malformed.                        |

> | AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = {"$id":"1","innerException":null,"message":"<ACTUAL ERROR MESSAGE>","typeName":"Microsoft.VisualStudio.Services.WebApi.VssInvalidPreviewVersionException, Microsoft.VisualStudio.Services.WebApi","typeKey":"VssInvalidPreviewVersionException","errorCode":0} | When the OIDC token request fails, the OIDC token api throws an error. More details about the specific error are specified in the "message" field of the Response as shown above. | Mitigation will usually depend on the scenario based on what [error message](https://learn.microsoft.com/azure/devops/pipelines/release/troubleshoot-workload-identity?view=azure-devops#error-messages) is being thrown. Make sure you use the [recommended Azure Pipelines task](https://learn.microsoft.com/azure/devops/pipelines/release/troubleshoot-workload-identity?view=azure-devops#review-pipeline-tasks). |

Reviewer 3: To help improve column width balance in the rendered table, I suggest breaking up the sample response by adding spaces between each value.
```suggestion
| AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = {"$id":"1", "innerException":null, "message":"`<ACTUAL ERROR MESSAGE>`", "typeName":"Microsoft.VisualStudio.Services.WebApi.VssInvalidPreviewVersionException, Microsoft.VisualStudio.Services.WebApi", "typeKey":"VssInvalidPreviewVersionException", "errorCode":0} | When the OIDC token request fails, the OIDC token api throws an error. More details about the specific error are specified in the "message" field of the Response as shown above. | Mitigation will usually depend on the scenario based on what [error message](https://learn.microsoft.com/azure/devops/pipelines/release/troubleshoot-workload-identity?view=azure-devops#error-messages) is being thrown. Make sure you use the [recommended Azure Pipelines task](https://learn.microsoft.com/azure/devops/pipelines/release/troubleshoot-workload-identity?view=azure-devops#review-pipeline-tasks). |
```

> | AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = Object moved to here. Status Code: 302.                                                                                                                                                                                                                        | The system access token seems to be malformed when passing in as a parameter to the credential.                                                                                   | System Access Token is a required System variable in the Azure Pipeline task and [should be provided in the pipeline task as mentioned in the docs](https://learn.microsoft.com/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#systemaccesstoken). Verify that the system access token value provided is the predefined variable in the Azure Devops and is not  malformed.                                                                                                                                                                         |

Reviewer: nit: spacing
```suggestion
| AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = Object moved to here. Status Code: 302.                                                                                                                                                                                                                        | The system access token seems to be malformed when passing in as a parameter to the credential.                                                                                   | System Access Token is a required System variable in the Azure Pipeline task and [should be provided in the pipeline task as mentioned in the docs](https://learn.microsoft.com/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#systemaccesstoken). Verify that the system access token value provided is the predefined variable in the Azure Devops and is not malformed.                                                                                                                                                                         |
```

> | AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = Object moved to here. Status Code: 302.                                                                                                                                                                                                                        | The system access token seems to be malformed when passing in as a parameter to the credential.                                                                                   | System Access Token is a required System variable in the Azure Pipeline task and [should be provided in the pipeline task as mentioned in the docs](https://learn.microsoft.com/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#systemaccesstoken). Verify that the system access token value provided is the predefined variable in the Azure Devops and is not  malformed.                                                                                                                                                                         |

Author: done

> | CredentialUnavailableError: AzurePipelinesCredential: is unavailable. Ensure that you're running this task in an Azure Pipeline, so that following missing system variable(s) can be defined- "SYSTEM_OIDCREQUESTURI"                                                                                                                                                    | This code is not running inside of the Azure Pipelines Environment. You may be running this code locally or on some other environment.                                            | This credential is only designed to run from inside the Azure Pipelines environment for the federated identity to work.                                                                                                                                                                                                                                                                                                |

Reviewer 2: ```suggestion
| CredentialUnavailableError: AzurePipelinesCredential: is unavailable. Ensure that you're running this task in an Azure Pipeline, so that following missing system variable(s) can be defined- "SYSTEM_OIDCREQUESTURI"                                                                                                                                                    | This code is not running inside of the Azure Pipelines environment. You may be running this code locally or on some other environment.                                            | This credential is only designed to run inside the Azure Pipelines environment for the federated identity to work.                                                                                                                                                                                                                                                                                                |
```

> | AuthenticationRequiredError: unauthorized_client: 700016 - AADSTS700016: Application with identifier 'clientId' was not found in the directory 'Microsoft'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.   | The `clientId` provided is invalid.                                                                                                                                               | Verify the client id argument is valid. If the service connection's federated identity was registered via a user-assigned managed identity, the client id of the managed identity should be provided. If the service connection's federated identity is registered via an app registration, the Application (client) ID from your app registration should be provided.                                                 |

Reviewer 2: ```suggestion
| AuthenticationRequiredError: unauthorized_client: 700016 - AADSTS700016: Application with identifier 'clientId' was not found in the directory 'Microsoft'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.   | The `clientId` provided is invalid.                                                                                                                                               | Verify the client ID argument is valid. If the service connection's federated identity was registered via a user-assigned managed identity, the client ID of the managed identity should be provided. If the service connection's federated identity is registered via an app registration, the Application (client) ID from your app registration should be provided.                                                 |
```

> | AuthenticationRequiredError: unauthorized_client: 700016 - AADSTS700016: Application with identifier 'clientId' was not found in the directory 'Microsoft'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.| The `clientId` provided is invalid. | Verify the client id argument is valid. If the service connection's federated identity was registered via a user-assigned managed identity, the client id of the managed identity should be provided. If the service connection's federated identity is registered via a Service Principal, the Application (client) ID from your app registration should be provided. |

Reviewer 4: I think the standard is to refer to the directory `'Contoso'` in docs?

Author: That might be for learn.microsoft. Within our TS guide, we've mostly used <'clientId'>

> | AuthenticationRequiredError: unauthorized_client: 700016 - AADSTS700016: Application with identifier 'clientId' was not found in the directory 'Microsoft'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.| The `clientId` provided is invalid. | Verify the client id argument is valid. If the service connection's federated identity was registered via a user-assigned managed identity, the client id of the managed identity should be provided. If the service connection's federated identity is registered via a Service Principal, the Application (client) ID from your app registration should be provided. |

Reviewer 4: Another case where we should say App Registration instead of Service Principal.

Author: done

| No service connection found with identifier <GUID>                                                                                                                                                                                                                                                                                                                       | The service connection id provided is incorrect.                                                                                                                                  | Verify the serviceConnectionId provided. This parameter refers to the `resourceId` of the Azure Service Connection. It can also be found in the queryString of the respective Service Connection's configuration page in Azure Devops.                                                                                                                                                                     |

> | something                                                                                                                                                                                                                                                                                                                                                                | somehting                                                                                                                                                                         | System Access Token is a required System variable in the Azure Pipeline task and [should be provided in the pipeline task as mentioned in the docs](https://learn.microsoft.com/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#systemaccesstoken).                                                                                                                                     |

Reviewer 5: There's an extra row here, probably a placeholder that needs to be removed - specifically, the row that just says "something"

> | AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = Object moved to here. Status Code: 302.                                                                                                                                                                                                                        | The system access token seems to be malformed when passing in as a parameter to the credential.                                                                                   | Verify that the system access token value provided is the predefined variable in the Azure Devops and is not being malformed.                                                                                                                                                                                                                                                                              |

Reviewer 5: ```suggestion
| AzurePipelinesCredential: Authentication Failed. oidcToken field not detected in the response. Response = Object moved to here. Status Code: 302.                                                                                                                                                                                                                        | The system access token seems to be malformed when passing in as a parameter to the credential.                                                                                   | Verify that the system access token value provided is the predefined variable in the Azure Devops and is not malformed.                                                                                                                                                                                                                                                                              |
```
